Bluejay/Bluejay/Error+Extensions.swift

```diff
@@ -0,0 +1,94 @@
+//
+//  Error+Extensions.swift
+//  Bluejay
+//
+//  Created by Ilya Laryionau on 16/09/2017.
+//  Copyright © 2017 Steamclock Software. All rights reserved.
+//
+
+import Foundation
+
+extension BJError: LocalizedError {
+    public var errorDescription: String? {
+        switch self {
+        case .bluetoothUnavailable:
+            return "Bluetooth unavailable."
+        case .multipleScanNotSupported:
+            return "Multiple scan is not supported."
+        case .multipleConnectNotSupported:
+            return "Multiple connect is not supported."
+        case .multipleDisconnectNotSupported:
+            return "Multiple disconnect is not supported."
+        case .connectionTimedOut:
+            return "Connection timed out."
+        case .notConnected:
+            return "Not connected to a peripheral."
+        case let .missingService(service):
+            return "Service not found: \(service.uuid)."
+        case let .missingCharacteristic(characteristic):
+            return "Characteristic not found: \(characteristic.uuid)."
+        case .cancelled:
+            return "Cancelled"
+        case .listenTimedOut:
+            return "Listen timed out."
+        case .readFailed:
+            return "Read failed."
+        case .writeFailed:
+            return "Write failed."
+        case .missingData:
+            return "No data from peripheral."
+        case let .dataOutOfBounds(start, length, count):
+            return "Cannot extract data with a size of \(count) using start: \(start), length: \(length)."
+        case let .unexpectedPeripheral(peripheral):
+            return "Unexpected peripheral: \(peripheral.uuid)."
+        case .scanningWithAllowDuplicatesInBackgroundNotSupported:
+            return "Scanning with allow duplicates while in the background is not supported."
+        case .missingServiceIdentifiersInBackground:
+            return "Scanning without specifying any service identifiers while in the background is not supported."
+        case .backgroundTaskRunning:
+            return "Regular Bluetooth operation is not available when a background task is running. For reading, writing, and listening, please use only the API found in the Synchronized Peripheral provided to you when working inside a background task block."
+        case .multipleBackgroundTaskNotSupported:
+            return "Multiple background task is not supported."
+        }
+    }
+}
+
+extension BJError: CustomNSError {
+    public var errorDomain: String {
+        return "Bluejay"
+    }
+
+    public var errorCode: Int {
```

Comment:
Think a better implementation of this is just to make the error an Int raw representable. That way this function can just be `return rawValue` and we can avoid having to maintain an exact numbering of the errors (just make bluetoothUnavailible == 1 and let the auto incrementing for Int enums numbers everything else)

Reply:
There're several reasons why I didn't make it `RawRepresentable`:
1. The protocol requires implementing `init?(rawValue: Self.RawValue)`. But we have cases with associated values, e.g. `missingService(ServiceIdentifier)`. So it's difficult and unclear how construct such values just as `Int`.
2. We cannot define it in the following way:
```
enum A: Int {
   case caseOne = 1
}
```
Again since associated values. Swift compiler won't allow us doing it.

So we'll have to create the same `switch` as for `errorCode`. But in case of `RawRepresentable` we'll have to implement `var rawValue` and `init`. I think this redundant.

Reply:
Ah yes, that makes sense.